torch/distributed/pipelining/schedules.py

```diff
@@ -1055,10 +1060,60 @@ def _get_pipeline_order(self) -> Optional[dict[int, list[Optional[_Action]]]]:
                     backward_mb += 1
                     remaining_backward -= 1
 
-            pipeline_order[rank] = actions
+            pipeline_order[rank] = _add_reduce_grad(actions)
         return pipeline_order
 
 
+def _requires_reduce_grad(action_type: _ComputationType) -> bool:
+    return action_type in (I, W, B)
+
+
+def _add_reduce_grad(actions: list[Optional[_Action]]) -> list[Optional[_Action]]:
+    """
+    REDUCE_GRAD refers to joint across minibatches grad reduction.
+    reduce_grad frees memory and we want to schedule it just after the last "backward"-like stage.
+    """
+    stage_with_bw_count: dict[int, int] = defaultdict(int)
```

Comment:
what's the point of counting, every stage should have the same number of backwards, equal to the number of microbatches, right?  is this just for validation sake?